{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done\n"
     ]
    }
   ],
   "source": [
    "from queries import start_elastic_search, index_documents\n",
    "from trec_files import make_trec_run\n",
    "from trec_files import read_eval_files\n",
    "from evaluation import trec_eval\n",
    "from evaluation import mean_average_precision\n",
    "from evaluation import print_trec_eval\n",
    "import elasticsearch\n",
    "import elasticsearch.helpers\n",
    "import json\n",
    "from elasticsearch import Elasticsearch\n",
    "import nltk\n",
    "from nltk.tokenize import word_tokenize\n",
    "from nltk.corpus import stopwords\n",
    "from nltk.stem import PorterStemmer\n",
    "from elasticsearch_dsl import Q\n",
    "from elasticsearch_dsl import Search\n",
    "\n",
    "print(\"done\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ut-student/.local/lib/python3.8/site-packages/elasticsearch/connection/base.py:200: ElasticsearchWarning: Elasticsearch built-in security features are not enabled. Without authentication, your cluster could be accessible to anyone. See https://www.elastic.co/guide/en/elasticsearch/reference/7.17/security-minimal-setup.html to enable security.\n",
      "  warnings.warn(message, category=ElasticsearchWarning)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(263080, [])"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Connect to the ElasticSearch server\n",
    "es = elasticsearch.Elasticsearch(host='localhost')  # in case you use Docker, the host is 'elasticsearch'\n",
    "\n",
    "# Index the collection into the index called 'genomics'\n",
    "body = {} # no indexing options (leave default)\n",
    "index_documents(es, 'data01/FIR-s05-medline.json', 'genomics-base', body)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\"cyclin-dependent kinase inhibitor 1A (p21, Cip1)\" in Homo sapiens \n",
      "\n",
      "\"cyclin-dependent kinase inhibitor 1A (p21, Cip1)\" in Homo sapiens tumor deficiency differential effects mice role age certain complex experimental \n",
      "\n",
      "\n",
      "ets variant gene 6 (TEL oncogene) in Homo sapiens \n",
      "\n",
      "ets variant gene 6 (TEL oncogene) in Homo sapiens phosphorylation map ser functional kinase regulation becomes bound constituent constitutive \n",
      "\n",
      "\n",
      "fibroblast growth factor 7 (keratinocyte growth factor) in Homo sapiens \n",
      "\n",
      "fibroblast growth factor 7 (keratinocyte growth factor) in Homo sapiens connective tissue activity factor growth wound around period suppression healing \n",
      "\n",
      "\n",
      "\"glycine receptor, alpha 1 (startle disease/hyperekplexia, stiff man syndrome)\" in Homo sapiens \n",
      "\n",
      "\"glycine receptor, alpha 1 (startle disease/hyperekplexia, stiff man syndrome)\" in Homo sapiens minor form major gene genetic defect generalized hereditary stiffness startle \n",
      "\n",
      "\n",
      "\"major histocompatibility complex, class II, DQ beta 1\" in Homo sapiens \n",
      "\n",
      "\"major histocompatibility complex, class II, DQ beta 1\" in Homo sapiens diabetes susceptibility type chief function related disease genetic protection resistance \n",
      "\n",
      "\n",
      "Janus kinase 2 (a protein tyrosine kinase) in Homo sapiens \n",
      "\n",
      "Janus kinase 2 (a protein tyrosine kinase) in Homo sapiens may enhancement kinase enhanced production synthesis tyrosine inhibitor camp epidermoid \n",
      "\n",
      "\n",
      "luteinizing hormone/choriogonadotropin receptor in Homo sapiens \n",
      "\n",
      "luteinizing hormone/choriogonadotropin receptor in Homo sapiens receptor regulate availability binding active cellular corpus desensitization factor fetal \n",
      "\n",
      "\n",
      "metallothionein 3 (growth inhibitory factor (neurotrophic)) in Homo sapiens \n",
      "\n",
      "metallothionein 3 (growth inhibitory factor (neurotrophic)) in Homo sapiens definition species human genetic ancient biological capacity considering contemporary criteria \n",
      "\n",
      "\n",
      "protein C (inactivator of coagulation factors Va and VIIIa) in Homo sapiens \n",
      "\n",
      "protein C (inactivator of coagulation factors Va and VIIIa) in Homo sapiens thrombin generation factor tissue surface protein source endothelial providing amount \n",
      "\n",
      "\n",
      "\"ras-related C3 botulinum toxin substrate 1 (rho family, small GTP binding protein Rac1)\" in Homo sapiens \n",
      "\n",
      "\"ras-related C3 botulinum toxin substrate 1 (rho family, small GTP binding protein Rac1)\" in Homo sapiens nuclear inhibitor translocation expression gene rho activation endothelial necrosis reduced \n",
      "\n",
      "\n",
      "tropomyosin 1 (alpha) in Homo sapiens \n",
      "\n",
      "tropomyosin 1 (alpha) in Homo sapiens definition human species genetic affect ancient capacity considering contemporary criteria \n",
      "\n",
      "\n",
      "frizzled homolog 4 (Drosophila) in Homo sapiens \n",
      "\n",
      "frizzled homolog 4 (Drosophila) in Homo sapiens degenerative differentiation entry identification meiotic mutant spermatid spermatocyte spermatogenesis albicans \n",
      "\n",
      "\n",
      "tyrosyl-tRNA synthetase in Homo sapiens \n",
      "\n",
      "tyrosyl-tRNA synthetase in Homo sapiens tyrosyl antibacterial aureus inhibitor activity action compound conclusion confirmation data \n",
      "\n",
      "\n",
      "major vault protein in Homo sapiens \n",
      "\n",
      "major vault protein in Homo sapiens definition species human genetic affect ancient biological capacity considering contemporary \n",
      "\n",
      "\n",
      "\"adrenergic receptor, alpha 1d\" in Mus musculus \n",
      "\n",
      "\"adrenergic receptor, alpha 1d\" in Mus musculus alpha receptor arterial blood cardiovascular directly function physiological pressure investigate \n",
      "\n",
      "\n",
      "\"caspase 3, apoptosis related cysteine protease\" in Mus musculus \n",
      "\n",
      "\"caspase 3, apoptosis related cysteine protease\" in Mus musculus cathepsin newt spermatogonial activity testis inhibitor involved chromosomal condensation culture \n",
      "\n",
      "\n",
      "\"tumor necrosis factor (ligand) superfamily, member 6\" in Mus musculus \n",
      "\n",
      "\"tumor necrosis factor (ligand) superfamily, member 6\" in Mus musculus neuronal cell affinity change classes inducer manipulate minimal peptide physiology \n",
      "\n",
      "\n",
      "Indian hedgehog in Mus musculus \n",
      "\n",
      "Indian hedgehog in Mus musculus genetic biochemical commensal considerable development extreme large novel potential profile \n",
      "\n",
      "\n",
      "\"phospholipase C, gamma 1\" in Mus musculus \n",
      "\n",
      "\"phospholipase C, gamma 1\" in Mus musculus mouse kappa overview gene locus number accession according available belonging \n",
      "\n",
      "\n",
      "seven in absentia 2 in Mus musculus \n",
      "\n",
      "seven in absentia 2 in Mus musculus genetic biochemical commensal considerable development extreme large novel potential presence \n",
      "\n",
      "\n",
      "\"deoxynucleotidyltransferase, terminal\" in Mus musculus \n",
      "\n",
      "\"deoxynucleotidyltransferase, terminal\" in Mus musculus genetic biochemical commensal considerable development extreme large novel potential presence \n",
      "\n",
      "\n",
      "\"cholinergic receptor, muscarinic 3\" in Rattus norvegicus \n",
      "\n",
      "\"cholinergic receptor, muscarinic 3\" in Rattus norvegicus mouse two frugivorous rat subspecies black chromosome diploid genomic karyotype \n",
      "\n",
      "\n",
      "Early growth response 1 in Rattus norvegicus \n",
      "\n",
      "Early growth response 1 in Rattus norvegicus mouse two frugivorous subspecies rat black chromosome diploid genomic karyotype \n",
      "\n",
      "\n",
      "Glucokinase in Rattus norvegicus \n",
      "\n",
      "Glucokinase in Rattus norvegicus mouse frugivorous two subspecies rat black chromosome diploid genomic karyotype \n",
      "\n",
      "\n",
      "\"Tachykinin (substance P, neurokinin A, neuropeptide K, neuropeptide gamma)\" in Rattus norvegicus \n",
      "\n",
      "\"Tachykinin (substance P, neurokinin A, neuropeptide K, neuropeptide gamma)\" in Rattus norvegicus protein production nuclear substance activation enhanced activity alone element factor \n",
      "\n",
      "\n",
      "estrogen receptor 1 in Rattus norvegicus \n",
      "\n",
      "estrogen receptor 1 in Rattus norvegicus mouse two frugivorous subspecies rat black chromosome diploid genomic karyotype \n",
      "\n",
      "\n",
      "cocaine and amphetamine regulated transcript in Rattus norvegicus \n",
      "\n",
      "cocaine and amphetamine regulated transcript in Rattus norvegicus mouse frugivorous two rat subspecies black chromosome diploid genomic karyotype \n",
      "\n",
      "\n",
      "hopscotch in Drosophila melanogaster \n",
      "\n",
      "hopscotch in Drosophila melanogaster pathway blood activation cell hop differentiation larval promote response hopscotch \n",
      "\n",
      "\n",
      "Slowpoke binding protein in Drosophila melanogaster \n",
      "\n",
      "Slowpoke binding protein in Drosophila melanogaster gene promoter amino expression region isolated sequence alignment approximately assay \n",
      "\n",
      "\n",
      "eiger in Drosophila melanogaster \n",
      "\n",
      "eiger in Drosophila melanogaster cell death pathway superfamily genetic protein although blocked demonstrate evidence \n",
      "\n",
      "\n",
      "shotgun in Drosophila melanogaster \n",
      "\n",
      "shotgun in Drosophila melanogaster gene amino promoter isolated protein expression region sequence acid alignment \n",
      "\n",
      "\n",
      "Signal-transducer and activator of transcription protein at 92E in Drosophila melanogaster \n",
      "\n",
      "Signal-transducer and activator of transcription protein at 92E in Drosophila melanogaster histamine protein kinase response tyrosine element melanoma squamous suppressed carcinoma \n",
      "\n",
      "\n",
      "crumbs in Drosophila melanogaster \n",
      "\n",
      "crumbs in Drosophila melanogaster za apical morphogenesis skeleton membrane necessary protein cell epithelia essential \n",
      "\n",
      "\n",
      "golgi phosphoprotein 2 in Homo sapiens \n",
      "\n",
      "golgi phosphoprotein 2 in Homo sapiens definition human species genetic affect ancient capacity considering contemporary criteria \n",
      "\n",
      "\n",
      "sorting nexin 9 in Homo sapiens \n",
      "\n",
      "sorting nexin 9 in Homo sapiens retrieval yeast different late snare classes coat continuously contrast distinct \n",
      "\n",
      "\n",
      "\"ankylosis, progressive homolog (mouse)\" in Homo sapiens \n",
      "\n",
      "\"ankylosis, progressive homolog (mouse)\" in Homo sapiens definition human species genetic affect ancient capacity considering contemporary criteria \n",
      "\n",
      "\n",
      "\"DNA cross-link repair 1C (PSO2 homolog, S. cerevisiae)\" in Homo sapiens \n",
      "\n",
      "\"DNA cross-link repair 1C (PSO2 homolog, S. cerevisiae)\" in Homo sapiens degenerative differentiation entry identification meiotic spermatid spermatocyte spermatogenesis mutant albicans \n",
      "\n",
      "\n",
      "Pax-8 protein in Rattus norvegicus \n",
      "\n",
      "Pax-8 protein in Rattus norvegicus mouse two frugivorous subspecies rat black chromosome diploid genomic karyotype \n",
      "\n",
      "\n",
      "done\n"
     ]
    }
   ],
   "source": [
    "# connect to ES server             \n",
    "es = elasticsearch.Elasticsearch('localhost')\n",
    "\n",
    "# Write the results of the queries contained in the topic file `'data/training-queries-simple.txt'` \n",
    "# to the run file `'baseline.run'`, and name this test as `test01`\n",
    "# make_trec_run(es, 'data01/FIR-s05-training-queries-simple.txt', 'baseline.run', \"genomics-base\", run_name='no_pseudo_01')\n",
    "make_trec_run(es, 'data01/FIR-s05-training-queries-simple.txt', 'baseline.run', \"genomics-base\", run_name='pseudo_01', pseudo_relevance=True)\n",
    "\n",
    "print(\"done\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Results for baseline.run\n",
      "mean success_at_1              0.1316\n",
      "mean success_at_5              0.2105\n",
      "mean success_at_10             0.3158\n",
      "mean r_precision               0.08062\n",
      "mean precision_at_1            0.1316\n",
      "mean precision_at_5            0.06842\n",
      "mean precision_at_10           0.05\n",
      "mean precision_at_50           0.01842\n",
      "mean precision_at_100          0.01105\n",
      "mean precision_at_recall_00    0.1984\n",
      "mean precision_at_recall_01    0.1923\n",
      "mean precision_at_recall_02    0.1467\n",
      "mean precision_at_recall_03    0.1308\n",
      "mean precision_at_recall_04    0.1239\n",
      "mean precision_at_recall_05    0.1183\n",
      "mean precision_at_recall_06    0.1054\n",
      "mean precision_at_recall_07    0.1036\n",
      "mean precision_at_recall_08    0.08429\n",
      "mean precision_at_recall_09    0.08381\n",
      "mean precision_at_recall_10    0.08381\n",
      "mean average_precision         0.1191\n"
     ]
    }
   ],
   "source": [
    "print_trec_eval('data01/FIR-s05-training-qrels.txt', 'baseline.run')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pseudo-Feedback Relevance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "initial_results = es.search(index='your_index_name', body=initial_query, size=10)\n",
    "\n",
    "# Extract the initial retrieved documents' content\n",
    "docs_content = [hit['_source']['field_to_extract'] for hit in initial_results['hits']['hits']]\n",
    "\n",
    "# Combine the initial query with terms from the top retrieved documents\n",
    "feedback_terms = []\n",
    "for doc_content in docs_content:\n",
    "    doc_tokens = word_tokenize(doc_content)\n",
    "    doc_tokens = [porter.stem(token) for token in doc_tokens if token.isalpha() and token not in stop_words]\n",
    "    feedback_terms.extend(doc_tokens)\n",
    "\n",
    "# Calculate term frequencies for the feedback terms\n",
    "feedback_term_freq = {}\n",
    "for term in feedback_terms:\n",
    "    if term in feedback_term_freq:\n",
    "        feedback_term_freq[term] += 1\n",
    "    else:\n",
    "        feedback_term_freq[term] = 1\n",
    "\n",
    "# Sort the feedback terms by their frequency\n",
    "sorted_feedback_terms = sorted(feedback_term_freq.items(), key=lambda x: x[1], reverse=True)\n",
    "\n",
    "# Choose the top feedback terms to expand the query\n",
    "num_feedback_terms = 5\n",
    "expanded_query = {\n",
    "    \"query\": {\n",
    "        \"match\": {\n",
    "            \"field_to_search\": {\n",
    "                \"query\": preprocess_text(\"your initial query\"),\n",
    "                \"operator\": \"or\"\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "}\n",
    "\n",
    "for term, _ in sorted_feedback_terms[:num_feedback_terms]:\n",
    "    expanded_query[\"query\"][\"match\"][\"field_to_search\"][\"query\"] += f\" {term}\"\n",
    "\n",
    "# Perform a new search with the expanded query\n",
    "new_results = es.search(index='your_index_name', body=expanded_query, size=10)\n",
    "\n",
    "# You can now work with the new search results in 'new_results'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Evaluate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "qrels_file = \"data01/trec.qrels\"\n",
    "run_file = \"data01/baseline.run\"\n",
    "\n",
    "# Read the relevant documents and retrieved documents\n",
    "all_relevant, all_retrieved = read_eval_files(qrels_file, run_file)\n",
    "\n",
    "# Evaluate the performance\n",
    "map_score = mean_average_precision(all_relevant, all_retrieved)\n",
    "print(f\"Mean Average Precision (MAP): {map_score}\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
